### PR TITLE
GH-3142: Add strict validation for unsigned types in ExampleParquetWriter

### LINKED ISSUE #3142: ExampleParquetWriter may write illegal values for uint_8 and uint_16
### Describe the bug, including details regarding any error messages, version, and platform.

The DataFusion Comet project's unit test use the ExampleParquetWriter to create Parquet files - https://github.com/apache/datafusion-comet/blob/996362e78d497c02542f1e29dbb7cba3ec16f64c/spark/src/test/scala/org/apache/spark/sql/CometTestBase.scala#L432
This was inspired by similar unit test code in Spark - https://github.com/apache/spark/blob/ece14704cc083f17689d2e0b9ab8e31cf71a7a2d/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala#L871

The output files can contain uint8 and uint16 values that are illegal per the spec. For example in this file - 

[alltypes_extended_plain.parquet.zip](https://github.com/user-attachments/files/18623383/alltypes_extended_plain.parquet.zip)

The columns `_8` and `_9` are `uint_8` and `uint_16` values and contain illegal negative values. 

```
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": true, "_2": 18, "_3": 10002, "_4": 10002, "_5": 10002, "_6": 10002.0, "_7": 10002.0, "_8": "100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002100021000210002", "_9": -18, "_10": -10002, "_11": -10002, "_12": -10002, "_13": "10002", "_14": [50, 50, 50], "_15": 10002, "_16": 10002, "_17": [50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50], "_18": 10002, "_19": 10002, "_20": 10002}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": true, "_2": 20, "_3": 10004, "_4": 10004, "_5": 10004, "_6": 10004.0, "_7": 10004.0, "_8": "100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004100041000410004", "_9": -20, "_10": -10004, "_11": -10004, "_12": -10004, "_13": "10004", "_14": [52, 52, 52], "_15": 10004, "_16": 10004, "_17": [52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52], "_18": 10004, "_19": 10004, "_20": 10004}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
{"_1": true, "_2": 24, "_3": 10008, "_4": 10008, "_5": 10008, "_6": 10008.0, "_7": 10008.0, "_8": "100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008100081000810008", "_9": -24, "_10": -10008, "_11": -10008, "_12": -10008, "_13": "10008", "_14": [56, 56, 56], "_15": 10008, "_16": 10008, "_17": [56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56], "_18": 10008, "_19": 10008, "_20": 10008}
{"_1": null, "_2": null, "_3": null, "_4": null, "_5": null, "_6": null, "_7": null, "_8": null, "_9": null, "_10": null, "_11": null, "_12": null, "_13": null, "_14": null, "_15": null, "_16": null, "_17": null, "_18": null, "_19": null, "_20": null}
```

Taking as an example the first value for column `_8` the bit pattern written to the file is `0xffffffee` which gets read as a negative value which is illegal for a unsigned int. 

The value originates in this line - https://github.com/apache/datafusion-comet/blob/996362e78d497c02542f1e29dbb7cba3ec16f64c/spark/src/test/scala/org/apache/spark/sql/CometTestBase.scala#L520 
where a negative value is cast to a byte and then written to Parquet.  The Parquet writer needs to cast correctly to a larger type before writing to the file. 

The values written can be read by the Parquet-java reader but other implementations are free to return an error or null for such values which is not desirable. 

### Component(s)

Core

### PR DESCRIPTION
**Pull request by ArnavBalyan**
Summary:
 - This PR addresses the issue of ExampleParquetWriter writing invalid values for unsigned integer types.
 - Updated the ExampleParquetWriter.Builder with optional withStrictUnsignedIntegerValidation which can be enabled by caller.
 - Fixes #3142

Testing:
 - Added UTs

### CODE REVIEW

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ExampleParquetWriter.java [hunk 1]
```diff
@@ -97,6 +97,7 @@ public static Builder builder(OutputFile file) {
 
   public static class Builder extends ParquetWriter.Builder<Group, Builder> {
     private MessageType type = null;
+    private boolean strictUnsignedIntegerValidation = false;
```
Comment (parthchandra) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ExampleParquetWriter.java]: I feel this should be true always. At least should be true by default.

Comment (ArnavBalyan) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ExampleParquetWriter.java]: Sure will will update if there are no other changes requested, cc @shangxinli @wgtmac could you please review thanks!

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/example/TestStrictUnsignedIntegerValidation.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.Types;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * Integration test for strict unsigned integer validation in ExampleParquetWriter.
+ */
+public class TestStrictUnsignedIntegerValidation {
```
Comment (shangxinli) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/example/TestStrictUnsignedIntegerValidation.java:44]: The test cases cover all unsigned types
  and both valid and invalid scenarios. Consider adding a test case for
  the maximum values of each unsigned type (255 for UINT_8, 65535 for 
  UINT_16, etc.) to ensure boundary conditions are properly handled.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  @Override
+  public void addDouble(double value) {
+    delegate.addDouble(value);
+  }
+
+  private void validateUnsignedInteger(int value) {
+    Type currentType = getCurrentFieldType();
+    if (currentType != null && currentType.isPrimitive()) {
+      LogicalTypeAnnotation logicalType = currentType.asPrimitiveType().getLogicalTypeAnnotation();
+      if (logicalType instanceof LogicalTypeAnnotation.IntLogicalTypeAnnotation) {
```
Comment (nandorKollar) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: Consider using `LogicalTypeAnnotationVisitor` here and in validateUnsignedLong method, instead of instanceof.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+/**
+ * A RecordConsumer that validates unsigned integer values to ensure they comply
+ * with the Parquet specification.
+ *
+ * This consumer wraps another RecordConsumer and validates that values written
+ * to unsigned integer fields (UINT_8, UINT_16, UINT_32, UINT_64) are within
+ * the valid unsigned range before delegating to the wrapped consumer.
+ */
+public class ValidatingUnsignedIntegerRecordConsumer extends RecordConsumer {
```
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: I'm not sure if it is enough to modify lines below to add the check:

https://github.com/apache/parquet-java/blob/master/parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java#L203-L206

WDYT? @parthchandra

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ExampleParquetWriter.java (hunk 1 above)
Comment (ArnavBalyan) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ExampleParquetWriter.java]: Updated this default value to true thanks!

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/example/TestStrictUnsignedIntegerValidation.java (hunk 2 above)
Comment (ArnavBalyan) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/example/TestStrictUnsignedIntegerValidation.java:44]: Updated thanks!

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java (hunk 3 above)
Comment (ArnavBalyan) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: Updated if you could PTAL

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java (hunk 4 above)
Comment (parthchandra) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: Hmm, that's a good idea. Keep in mind that we will get the validation in any place where a writer is created withValidation, not just in the example. I'm in favor of it though.

Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: Yes, that is consistent with other validations. I just think that adding a new class for unsigned integer is an overkill, especially when we already have a ValidatingRecordConsumer.java that serves the same purpose.

Comment (ArnavBalyan) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/example/ValidatingUnsignedIntegerRecordConsumer.java]: Yes sure, will update to piggyback on ValidatingRecordConsumer

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java [hunk 5]
```diff
@@ -46,7 +48,11 @@
 public class ValidatingRecordConsumer extends RecordConsumer {
   private static final Logger LOG = LoggerFactory.getLogger(ValidatingRecordConsumer.class);
 
+  private static final int UINT_8_MAX_VALUE = 255;
+  private static final int UINT_16_MAX_VALUE = 65535;
+
   private final RecordConsumer delegate;
+  private final boolean strictUnsignedIntegerValidation;
```
Comment (wgtmac) [inline parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:55]: Shouldn't it be validated at all times? This is already the ValidatingRecordConsumer so it might need to validate everything.

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java [hunk 6]
```diff
@@ -46,7 +48,11 @@
 public class ValidatingRecordConsumer extends RecordConsumer {
   private static final Logger LOG = LoggerFactory.getLogger(ValidatingRecordConsumer.class);
 
+  private static final int UINT_8_MAX_VALUE = 255;
+  private static final int UINT_16_MAX_VALUE = 65535;
```
Comment (wgtmac) [inline parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:52]: Why only 8 and 16?

Comment (parthchandra) [reply parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:52]: UINT_8, UINT_16, and UINT_32 annotate an int32 value, UINT_64 annotatyes a i64 value. For UINT_8, and UINT16 valid values are in the range [0, UINT_8_MAX_VALUE] and [0, UINT_16_MAX_VALUE] respectively. For UINT_32 and UINT_64, all non negative values are valid.

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java (hunk 5 above)
Comment (parthchandra) [reply parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:55]: +1

**Review by parthchandra — COMMENTED** [review]
lgtm

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java (hunk 5 above)
Comment (ArnavBalyan) [reply parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:55]: Unfortunately there are several test breaking and existing code built with the assumption of no validation. Since we put the changes directly within `ValidatingRecordConsumer` it causes existing behaviour to break. I added this with a feature flag so new users can start using the validation. I'll investigate as a followup if we can cleanup/fix existing code thanks!

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java (hunk 6 above)
Comment (ArnavBalyan) [reply parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:52]: yes that was the reason thanks!

File: parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java (hunk 5 above)
Comment (wgtmac) [reply parquet-column/src/main/java/org/apache/parquet/io/ValidatingRecordConsumer.java:55]: Could you help investigate why existing test break? Are all of them failed because of unsigned integer overflow? If yes, they are actually bugs and should be fixed in this PR so we don't have to add a separate flag for unsigned integer.
